### PR TITLE
Optional mandatory password

### PR DESCRIPTION
**Pull request by sesquipedalian-dev**
Make (configurably) passwords mandatory for uploads and implement password requirements checking.

- add values to config to make passwords mandatory, and to provide a list of password validation rules by name
- Updated download on server to abandon early (400) when PASSWORD_REQUIRED is true if the file was somehow saved without a password. This should backstop attempts to modify the frontend code to not assign a password, since the file will ultimately not be downloadable.
- Update frontend to make the password field always appear if PASSWORD_REQUIRED is true
- updated frontend so that upload button changes color and doesn't do anything, if a password is required and the password doesn't meet the parameters
- Fixed password message to allow multiline text, for multiple error cases
- added more password validation options, such as by character classes and minimum length

Tested by running like this: `PASSWORD_REQUIREMENTS_LIST=min_24,number_1,special_1 PASSWORD_REQUIRED=true npm start ` and putting in a password that doesn't meet the requirements. 

Screen Shot 2019-12-04 at 2 40 03 PM

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -5,27 +5,6 @@
 ##
 
 
-# Build project
```
Comment (zhouzhuojie) [inline Dockerfile]: oh, is it the reason we checked in the `dist` folder?

Comment (zhouzhuojie) [reply Dockerfile]: I think we should still keep the building step, it keeps the repo clean and at the same time make sure the dist folder is always generated from the source, what do you think?

Comment (sesquipedalian-dev) [reply Dockerfile]: not sure, I think I picked this up while rebasing. @cjford ?

Comment (cjford) [reply Dockerfile]: Yeah this was a temporary experiment, after you branched it was removed via hard reset. If you rebase again on current origin/master it should go away.

File: app/passwordValidator.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        classDescription,
+        count
+      });
+    }
+    return null;
+  };
+}
+
+let validationRules = {
+  length_32: maxLengthValidate(32),
```
Comment (cjford) [inline app/passwordValidator.js]: Works for our use-case, but if this is going upstream shouldn't configuration support arbitrary numbers for each rule? I think node-convict could support this with a custom or object format.

Comment (sesquipedalian-dev) [reply app/passwordValidator.js]: Hm, @gusk also suggested something like this one: https://validatejs.org/ .  Looking at that it seems kind of complicated; I'd like it to still be configurable with a simple environment variable.  I think you're right that the current implementation is too rigid.

Comment (cjford) [reply app/passwordValidator.js]: What if we split each rule string on `_` and `parseInt` the second element?

Comment (sesquipedalian-dev) [reply app/passwordValidator.js]: I added a nested object to convict. I couldn't figure out how to get convict to parse an object from an environment variable, so I added a config.json file to put in this and any other config.  Please have another look @cjford :)

**Review by charli-watson — APPROVED** [review]

File: server/config.js [hunk 3]
```diff
@@ -3,7 +3,73 @@ const { tmpdir } = require('os');
...
+      format: Number,
+      default: 32
+    }
+  },
+  minLength: {
+    type: {
+      format: ['minLength']
+    },
+    length: {
+      doc: 'the maximum length allowed for passwords',
```
Comment (oriahu) [inline server/config.js:22]: noticed this documentation is the same as for maxLength.
also, the max length is set to the min length
are these the base default configuration?

Comment (sesquipedalian-dev) [reply server/config.js:22]: fixed in master

**Pull request merged**